### PR TITLE
CI Tekton Migration to Github Action

### PR DESCRIPTION
**Pull request by yasserisa**
* Tekton Pipelines have been migrated to Github Action, with 100% support for tier 1 operating systems.
  * RPC tests are not included in this PR, and will be in a follow-up PR.

* Adjustments were made based on the PR: [https://github.com/zcash/zcash/pull/6804](https://github.com/zcash/zcash/pull/6804) to ensure compatibility of the `sec-hard test` with Github Action.

* The `full_test_suite.py` file has been modified to provide support for the Windows architecture (`x86_64-w64-mingw32`).

* Some tests are not functioning on Windows, so they have been deactivated to save time and costs.

* It was left pending for a future PR to make it compatible with workflow_dispatch. This enables more granular execution based on variables, allowing, for example, the execution of a specific tier or environment.

### CODE REVIEW

**Comment by str4d** [comment]
[Force-pushed](https://github.com/zcash/zcash/compare/ac1b5366b9d914da9c94bd78bd0c2f2bf3610af8..8e9e2f07a5ac31d712a051d0227773ab26dd8c1d) to clean up the changes and split some of them out into separate commits, ahead of rebasing.

File: .github/workflows/ci.yml
Comment (str4d) [inline .github/workflows/ci.yml:147]: Is this useful to set for the cached build dependencies (and similarly on the CCache files)? AFAICT there shouldn't be anything that can be shared between OSs, because they all compile for different targets. Note that we include `matrix.name` in the cache key, so they won't be used cross-OS anyway (unlike the Sprout parameters for which this is set in the next commit).

File: .github/workflows/ci.yml
Comment (str4d) [inline .github/workflows/ci.yml:353]: Rather than partially reconstructing the path for the uploaded artifact, we should just use [the output that the other job provides](https://github.com/google-github-actions/upload-cloud-storage?tab=readme-ov-file#outputs). In this case it would be something like `${{ needs.build.steps.upload-btest.outputs.uploaded }}`.

File: .github/gcs-download-artifacts.sh
Comment (str4d) [inline .github/gcs-download-artifacts.sh:1]: This would be more usable as a composite action (so we aren't beholden to Bash CLI parsing issues). See [this example in the Rust crate repo](https://github.com/zcash/librustzcash/blob/3cbc4815b6785aa53f48e5f61acbcfb18adda59a/.github/actions/prepare/action.yml) ([used here](https://github.com/zcash/librustzcash/blob/3cbc4815b6785aa53f48e5f61acbcfb18adda59a/.github/workflows/ci.yml#L34-L37)).

File: .github/workflows/ci.yml
Comment (str4d) [inline .github/workflows/ci.yml:343]: It's not clear to me why you are checking out the repo in this job; the Boost test runner does not require anything other than the test binary.

File: .github/workflows/ci.yml
Comment (str4d) [inline .github/workflows/ci.yml:376]: Ditto; the gtest runner does not require anything other than the test binary.

File: .github/workflows/ci.yml
Comment (str4d) [inline .github/workflows/ci.yml:429]: Rather than disabling every individual step, we should have a separate `test_matrix` subset (maybe named `unix_test_matrix`) that excludes Windows.

File: .github/workflows/ci.yml
Comment (str4d) [inline .github/workflows/ci.yml:696]: Likewise, move this configuration into `rpc_test_matrix` instead of having it on every step (which is a maintenance hazard).

File: qa/zcash/full_test_suite.py
Comment (str4d) [inline qa/zcash/full_test_suite.py:117]: Why was this changed to not use `repofile`? `full_test_suite.py` uses that to make itself independent of the directory in which it is run. The other new path additions should be using it as well.

**Comment by str4d** [comment]
There are a bunch of identical warnings:
`The "process_gcloudignore" option is true, but no .gcloudignore file was found. If you do not intend to process a gcloudignore file, set "process_gcloudignore" to false.`

File: qa/zcash/full_test_suite.py
Comment (yasserisa) [reply qa/zcash/full_test_suite.py:117]: Good observation, it's better to use `repofile`. I think I accidentally removed that while running a unit test. I'll revert that in the next commit

File: .github/workflows/ci.yml
Comment (yasserisa) [reply .github/workflows/ci.yml:343]: Indeed, it's not necessary. Good catch. The fix will be included in the next commit

File: .github/workflows/ci.yml
Comment (yasserisa) [reply .github/workflows/ci.yml:353]: I like the idea, I'm going to try how it works

File: .github/workflows/ci.yml
Comment (yasserisa) [reply .github/workflows/ci.yml:376]: Indeed, it's not necessary. Good catch. The fix will be included in the next commit

File: .github/workflows/ci.yml
Comment (yasserisa) [reply .github/workflows/ci.yml:429]: Yes, that could be a good workaround, thus avoiding several if statements

File: .github/workflows/ci.yml
Comment (yasserisa) [reply .github/workflows/ci.yml:696]: Yes, it could go in the setup step. I designed it that way to avoid adding more minutes to the first step without knowing if the previous tests/builds would be successful and to avoid spending more minutes

File: .github/workflows/ci.yml
Comment (yasserisa) [reply .github/workflows/ci.yml:147]: As you mention, it's not necessary. It was my mistake to copy from the example of the Sprout cache code. The fix will be included in the next commit

Comment (str4d) [reply .github/workflows/ci.yml:147]: No need to include a fix; I extracted this change into a separate commit, so you can just rebase and drop that commit.

File: .github/gcs-download-artifacts.sh
Comment (yasserisa) [reply .github/gcs-download-artifacts.sh:1]: https://github.com/zcash/zcash/pull/6820/commits/18d7fe9551de6894604ff141de6588ca0d663f22

File: .github/workflows/ci.yml
Comment (yasserisa) [reply .github/workflows/ci.yml]: It seems that implementing this complicates the code and its potential maintenance.

File: .github/workflows/ci.yml
Comment (yasserisa) [reply .github/workflows/ci.yml:339]: It is added again due to the composite action

File: .github/workflows/ci.yml
Comment (yasserisa) [reply .github/workflows/ci.yml:373]: It is added again due to the composite action

**Comment by str4d** [comment]
[Force-pushed](https://github.com/zcash/zcash/compare/7ee4f73d62ecfac2279e8516934906e1534e5102..f68b3fe5f6b9dc8c7ebfd671047e32fe1da13986) to:
- Finish replacing the Bash script with a composite Action.
- Replace per-step Windows exclusion with a Unix matrix.
  - This is more efficient as it doesn't even start the Windows runner for those jobs, and thus doesn't count towards our active job limit.
- Fix the `full_test_suite.py` changes to be path-independent.
- Run the Rust tests directly (using `rust-toolchain.toml` for version pinning) instead of using the versioned Rust compiler.
- Split up the remaining changes into separate commits and clean a few things up.

**Comment by str4d** [comment]
[Force-pushed](https://github.com/zcash/zcash/compare/f68b3fe5f6b9dc8c7ebfd671047e32fe1da13986..6d5bc8af296c60ee70878cce66fb1f9af50f13fb) to fix syntax errors.

Now CI is failing with a GCS authentication error:
```
Run google-github-actions/auth@v2
  with:
    credentials_json: ***
    create_credentials_file: true
    export_environment_variables: true
    universe: googleapis.com
    cleanup_credentials: true
    access_token_lifetime: 3600s
    access_token_scopes: https://www.googleapis.com/auth/cloud-platform
    id_token_include_email: false
Created credentials file at "/home/runner/work/zcash/zcash/gha-creds-8b35c58f0c02a76e.json"

[..]

Run gsutil -m cp -r gs://gh-zcash/8120135108/ubuntu-20.04/src/test/test_bitcoin /home/runner/work/_temp/gcs-download
  gsutil -m cp -r gs://gh-zcash/8120135108/ubuntu-20.04/src/test/test_bitcoin /home/runner/work/_temp/gcs-download
  shell: /usr/bin/bash --noprofile --norc -e -o pipefail {0}
  env:
    CLOUDSDK_AUTH_CREDENTIAL_FILE_OVERRIDE: /home/runner/work/zcash/zcash/gha-creds-8b35c58f0c02a76e.json
    GOOGLE_APPLICATION_CREDENTIALS: /home/runner/work/zcash/zcash/gha-creds-8b35c58f0c02a76e.json
    GOOGLE_GHA_CREDS_PATH: /home/runner/work/zcash/zcash/gha-creds-8b35c58f0c02a76e.json
    CLOUDSDK_CORE_PROJECT: ecc-infra-prod
    CLOUDSDK_PROJECT: ecc-infra-prod
    GCLOUD_PROJECT: ecc-infra-prod
    GCP_PROJECT: ecc-infra-prod
    GOOGLE_CLOUD_PROJECT: ecc-infra-prod
ServiceException: 401 Anonymous caller does not have storage.objects.get access to the Google Cloud Storage object. Permission 'storage.objects.get' denied on resource (or it may not exist).
```

This will be due to me switching from:
```bash
# $1 = ${{ secrets.GCP_SA_KEY }}
echo $1 | base64 -d > json.json
gcloud auth activate-service-account --key-file=json.json
```
inside the old Bash script, to:
```yaml
      - name: Authenticate to Google Cloud
        uses: google-github-actions/auth@v2
        with:
          credentials_json: ${{ secrets.GCP_SA_KEY }}
```
in the main CI file before calling the composite Action.

**Comment by str4d** [comment]
I noticed that [the README for the `auth` action](https://github.com/google-github-actions/auth?tab=readme-ov-file#prerequisites) said it requires Cloud SDK 390+, and CI was using 363+, so just in case that's the problem [I've bumped the version requirement](https://github.com/zcash/zcash/compare/6d5bc8af296c60ee70878cce66fb1f9af50f13fb..bbe22755c2ebf69d928b4c818ec55f92f478d4e5).

File: .github/workflows/ci.yml
Comment (str4d) [inline .github/workflows/ci.yml:707]: I want to solve this before we merge the RPC tests into Actions CI. Right now, because `needs.setup.outputs.rpc_tests` is a list of RPC test names, the effect of combining `name` with `rpc_tests` is that we create a separate job per RPC test per runner, which I don't believe is actually faster than batching and parallelizing within a runner (particularly because there is an upper limit on the number of jobs we can run at once).

Instead, I think we need to change `needs.setup.outputs.rpc_tests` to be a list of lists of RPC test names, with the code in `setup` deciding how to split the tests into different matrix entry groups.

To avoid blocking this PR, I propose that once we get everything else working, we move the last commit (that adds the RPC tests) to a separate branch, remove it from this one, and then merge this PR to get the rest of the changes in (which we can then remove from Tekton CI).

File: .github/workflows/ci.yml
Comment (daira) [inline .github/workflows/ci.yml:785]: Why is it necessary to do this given that [qa/pull-tester/rpc-tests.py](https://github.com/zcash/zcash/blob/master/qa/pull-tester/rpc-tests.py) already takes a list of tests as a command-line argument?

(It is okay to modify that script if needed.)

**Review by daira — APPROVED** [review]
utACK modulo @str4d's comment (https://github.com/zcash/zcash/pull/6820#discussion_r1509884346).

File: .github/workflows/ci.yml
Comment (yasserisa) [reply .github/workflows/ci.yml:707]: To parallelize within a single runner, would we do it within the Python script?

**Comment by yasserisa** [comment]
> [Force-pushed](https://github.com/zcash/zcash/compare/f68b3fe5f6b9dc8c7ebfd671047e32fe1da13986..6d5bc8af296c60ee70878cce66fb1f9af50f13fb) to fix syntax errors.
> 
> Now CI is failing with a GCS authentication error:
> 
> ```
> Run google-github-actions/auth@v2
>   with:
>     credentials_json: ***
>     create_credentials_file: true
>     export_environment_variables: true
>     universe: googleapis.com
>     cleanup_credentials: true
>     access_token_lifetime: 3600s
>     access_token_scopes: https://www.googleapis.com/auth/cloud-platform
>     id_token_include_email: false
> Created credentials file at "/home/runner/work/zcash/zcash/gha-creds-8b35c58f0c02a76e.json"
> 
> [..]
> 
> Run gsutil -m cp -r gs://gh-zcash/8120135108/ubuntu-20.04/src/test/test_bitcoin /home/runner/work/_temp/gcs-download
>   gsutil -m cp -r gs://gh-zcash/8120135108/ubuntu-20.04/src/test/test_bitcoin /home/runner/work/_temp/gcs-download
>   shell: /usr/bin/bash --noprofile --norc -e -o pipefail {0}
>   env:
>     CLOUDSDK_AUTH_CREDENTIAL_FILE_OVERRIDE: /home/runner/work/zcash/zcash/gha-creds-8b35c58f0c02a76e.json
>     GOOGLE_APPLICATION_CREDENTIALS: /home/runner/work/zcash/zcash/gha-creds-8b35c58f0c02a76e.json
>     GOOGLE_GHA_CREDS_PATH: /home/runner/work/zcash/zcash/gha-creds-8b35c58f0c02a76e.json
>     CLOUDSDK_CORE_PROJECT: ecc-infra-prod
>     CLOUDSDK_PROJECT: ecc-infra-prod
>     GCLOUD_PROJECT: ecc-infra-prod
>     GCP_PROJECT: ecc-infra-prod
>     GOOGLE_CLOUD_PROJECT: ecc-infra-prod
> ServiceException: 401 Anonymous caller does not have storage.objects.get access to the Google Cloud Storage object. Permission 'storage.objects.get' denied on resource (or it may not exist).
> ```
> 
> This will be due to me switching from:
> 
> ```shell
> # $1 = ${{ secrets.GCP_SA_KEY }}
> echo $1 | base64 -d > json.json
> gcloud auth activate-service-account --key-file=json.json
> ```
> 
> inside the old Bash script, to:
> 
> ```yaml
>       - name: Authenticate to Google Cloud
>         uses: google-github-actions/auth@v2
>         with:
>           credentials_json: ${{ secrets.GCP_SA_KEY }}
> ```
> 
> in the main CI file before calling the composite Action.

I had the same problem, that's why I added that authentication step to the script. An alternative workaround is to move that same command to one of the action's steps and remove the authentication one from the SDK

File: .github/workflows/ci.yml
Comment (yasserisa) [reply .github/workflows/ci.yml:785]: This was the workaround to run the tests in parallel in Tekton, since the list of tests is hardcoded in 3 variables within that script. At that time, it was decided not to modify the logic of the script and continue maintaining the list within that script

**Comment by str4d** [comment]
[Force-pushed](https://github.com/zcash/zcash/compare/bbe22755c2ebf69d928b4c818ec55f92f478d4e5..db2320ca28f02263f1e17fe28b983a65098e777c) to switch from `gsutil` to `gcloud storage` for downloading artifacts, as the latter seems to work with `google-github-actions/auth@v2` per @yasserisa's testing.

**Comment by str4d** [comment]
Okay, there are three more remaining classes of errors to fix.

- On Windows runners, there's some kind of issue where backslashes in Windows paths are being treated as Bash escapes:
```
$ rm -rf D:\a\_temp/gcs-download
$ mkdir D:\a\_temp/gcs-download
mkdir: cannot create directory 'D:a_temp/gcs-download': No such file or directory
```

- The `secp256k1` and `univalue` tests are failing because their artifacts are a directory, and due to us checking out the source repository, the target directories already exist.
- The Rust tests are failing on Windows with a linker error, which is probably also a path issue (but this isn't a blocker for the PR, we can fix it later).

**Comment by str4d** [comment]
[Force-pushed](https://github.com/zcash/zcash/compare/db2320ca28f02263f1e17fe28b983a65098e777c..d4d685205f082ba4c0a74793e2fa19b4bb7f8809) to fix the first two issues:
- We now remove `src/secp256k1` and `src/univalue` before moving the artifacts into place; this works because the artifacts are scoped to the particular workflow run, so we can use dead reckoning to ensure the deleted source files line up with the artifact ones.
- We switch from `${{ runner.temp }}` to `$RUNNER_TEMP` in an attempt to avoid the escaping problem (as suggested in https://github.com/orgs/community/discussions/25910#discussioncomment-3249642).

The RPC tests are also removed from the PR; they will be added back in a separate subsequent PR.

**Comment by str4d** [comment]
[Force-pushed](https://github.com/zcash/zcash/compare/d4d685205f082ba4c0a74793e2fa19b4bb7f8809..08f115da8a59f2c37cf341ffe476fbc24316e88c) to fix the last CI-level Windows oddity. The remaining failing tests are Windows-specific issues affecting a couple of individual tests, that don't need to block this PR merging.

**Review by str4d — APPROVED** [review]
ACK 08f115da8a59f2c37cf341ffe476fbc24316e88c. One of the remaining test failures (Boost tests on Windows) is pre-existing, and the other (Rust tests on Windows) can be addressed in a subsequent PR.

**Pull request merged**
